man/man8/puppetmasterd.8
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "PUPPET\-MASTER" "8" "June 2012" "Puppet Labs, LLC" "Puppet manual"
.
.SH "NAME"
\fBpuppet\-master\fR \- The puppet master daemon
.
.SH "SYNOPSIS"
The central puppet server\. Functions as a certificate authority by default\.
.
.SH "USAGE"
puppet master [\-D|\-\-daemonize|\-\-no\-daemonize] [\-d|\-\-debug] [\-h|\-\-help] [\-l|\-\-logdest \fIfile\fR|console|syslog] [\-v|\-\-verbose] [\-V|\-\-version] [\-\-compile \fInode\-name\fR]
.
.SH "DESCRIPTION"
This command starts an instance of puppet master, running as a daemon and using Ruby\'s built\-in Webrick webserver\. Puppet master can also be managed by other application servers; when this is the case, this executable is not used\.
.
.SH "OPTIONS"
Note that any configuration parameter that\'s valid in the configuration file is also a valid long argument\. For example, \'ssldir\' is a valid configuration parameter, so you can specify \'\-\-ssldir \fIdirectory\fR\' as an argument\.
.
.P
See the configuration file documentation at http://docs\.puppetlabs\.com/references/stable/configuration\.html for the full list of acceptable parameters\. A commented list of all configuration options can also be generated by running puppet master with \'\-\-genconfig\'\.
.
.TP
\-\-daemonize
Send the process into the background\. This is the default\.
.
.TP
\-\-no\-daemonize
Do not send the process into the background\.
.
.TP
\-\-debug
Enable full debugging\.
.
.TP
\-\-help
Print this help message\.
.
.TP
\-\-logdest
Where to send messages\. Choose between syslog, the console, and a log file\. Defaults to sending messages to syslog, or the console if debugging or verbosity is enabled\.
.
.TP
\-\-verbose
Enable verbosity\.
.
.TP
\-\-version
Print the puppet version number and exit\.
.
.TP
\-\-compile
Compile a catalogue and output it in JSON from the puppet master\. Uses facts contained in the $vardir/yaml/ directory to compile the catalog\.
.
.SH "EXAMPLE"
puppet master
.
.SH "DIAGNOSTICS"
When running as a standalone daemon, puppet master accepts the following signals:
.
.TP
SIGHUP
Restart the puppet master server\.
.
.TP
SIGINT and SIGTERM
Shut down the puppet master server\.
.
.SH "AUTHOR"
Luke Kanies
.
.SH "COPYRIGHT"
Copyright (c) 2011 Puppet Labs, LLC Licensed under the Apache 2\.0 License


CHANGELOG
3.0.0rc3
===
e619304 (#14761) Add boot, reboot to excludes list for redhat provider
7c1d8c2 (maint) Disable upstart spec test on windows
37df5c4 (#14615) Final stub fixes for the tests.
f11dfd6 (#14615) Fix provider spec tests to work on non gentoo systems
060810e maint: Make it easier to test the service provider
c0cb0fa (Maint) Close filehandles after read in upstart
55505db (Maint) Fix upstart provider to work on Ruby 1.9.3
b0f1d58 Update redhat.spec dependency to include hiera-puppet
6e2f23d Update puppet gem dependencies
d19f93f Update redhat init scripts to use `puppet (agent|queue|master)`
a45942b (#14297) Don't remove tty1 or rsyslog during upstart acceptance
466ab49 (#14297) Define agent in upstart acceptance test
f5d1f74 (#14297) Update upstart provider acceptance tests
3c5ef40 (#14297) Clean up upstart provider and tests
36c41e1 (#14343) Test cases for more forms of variable lookup
6cca061 (#14297) Add acceptance test for upstart provider
b9ed026 (#14297) Update spec tests for upstart provider
91827e8 (#14297) Handle upstart services better
3541db6 (#14343) Lookup in correct inherited scope
2d83e22 (Maint) Added test for fully qualified variables
0788ea2 (#14615) Exclude helperscripts in gentoo service provider
2864d86 (maint) Change Facter dep from exit to raise
2ec9b74 (maint) Make Puppet have hard dependency on Facter 2.0
6c37538 (#3909) Move default certname logic to Puppet::Settings
6be09f0 (#3909) Strip trailing dot from default FQDN

3.0.0rc2
===
0cea47e (#14609) Fix master initialization when running under Passenger
b6c5e9d (#14514) Check for the presence of hiera config
ca5271f (maint) Add data binding Hiera tests

3.0.0rc1
===
a731e15 Update CHANGELOG lib/puppet.rb puppet.spec for 2.7.15rc1
53e0057 Use internal domain name for module tool tests
f4ba59e Use absolute_path in module install spec
ead9d25 (Maint) Color console output when the win32console gem is installed
a76d873 (#14424) Expand path of the target directory
95b75e1 Fix unit test failure in fileserver due to modules
da5361d Maint: De-genericize Puppet Forge references in module face
3b144bf Revert "maint: use relative paths to fixtures."
dae5539 (#14440) Hack to allow help faces to be used as API
e862cad Fix broken man page rake task
b26699a (#10146) `-` is not legal in variable names.
7ce9e17 maint: use relative paths to fixtures.
a7a532c Significantly improve compilation performance when using modules
12c5bda Fix a stub of ansicolor that I missed
ea740f6 Add Puppet::Util::Platform to abstract platform checks
91a4a7d Don't use features to check for color in logging
80fe59d Remove a bad test that can lead to random hangs...
0efb4c2 (#14296) Improve template function error message.
d5e0cdd (#14288) gem provider should tolerate bad lines.
574e4c8 Update CHANGELOG, spec files for 2.7.14
d2b3ba0 Revert "Fix incorrect argument handling for expire in NodeExpirer"
7c7d4b6 Fix #14123 for Windows
1c9bec0 Fix incorrect argument handling for expire in NodeExpirer
7e92a7c Fix filebucket specs on Windows
a713f07 (#8778) Make '' == undef commutative in the DSL
24d4bb5 Revert "(#11004) Fix regex for pkg provider line matching"
34da6f7 (#13651) Use magenta, not purple in Facter::Util::Colors
e37b536 (#11004) Fix regex for pkg provider line matching
9b47110 (#13956) Fix failing spec tests
7e92dcb Just change Puppet settings in tests, rather than stubbing.
ca313dc (maint) Fix destinations_spec failure on windows
cde3bee Just change Puppet settings in tests, rather than stubbing.
6aa8441 (#11004) Support Solaris 11 output in pkg package provider
61e7d3c maint: Fix pkg_spec test
603b368 (#14173) Enforce that filebucket paths must be absolute
b7828ed (#3452) Autorequire user for cron
a47c42d Update CHANGELOG for 2.7.14rc3
402a425 (#14127) Add integration tests for ssh_authorized_key
85ee441 maint: refactor integration specs for ssh_authorized_key
42b20be (#13682) Fix acceptance test failures
dce3c32 (#13682) Rename Puppet::Module::Tool to Puppet::ModuleTool
5117e12 (maint) Ensure every file has a trailing newlines
b4d1c65 (#14127) ssh_authorized_keys grammer fails on blank lines.
25ad6d7 Updating CHANGELOG for Puppet 2.7.14rc2
74e462f (#14060) Fix quoting of commands to interpolate inside the shell.
796a072 Modernize the style of the shell exec provider tests.
9e526c9 (#14060) Fix quoting of commands to interpolate inside the shell.
36d959a (#14123) Puppet shouldn't explode if PATH contains ~nonexistent_user
a009c54 Modernize the style of the shell exec provider tests.
934ec89 Improve deprecation warning for dynamic lookup
c255414 (#14036) Add upstart tests
dc3bb9f (#14036) Handle upstart better
a281c48 (#13645) Add test cases to tagmail_spec
3a4ac60 Updating CHANGELOG, lib/puppet.rb for 2.7.14rc1
f6ac1bb (#13645) Do not open a smtp connection if nothing to report
f40cc71 (#12392): Colorize console output on Windows
0d9e852 Removed duplication of compiling a catalog
d67711d Cleaner test for create_resources doing nothing
2d3cc63 Fixing problem caused by activerecord
f1331ba Put back require 'puppet' in Puppet::Util
79bf4fa (#13299) Fix error message when agent is disabled
5178adc Update puppet.spec for rpm packaging
fb96747 Fixed old log test to match new autoflush behavior
878127f Make sure the log file writable
72d3b56 Default autoflushing of log files to true
9711d50 Created test that shows enc warning problem
d54100d (Maint) Fix order-dependent test failure
f531400 (Maint) Don't assume eventlog gem is installed on Windows
2957ffe Fix tempfile cleanup in specs
e2d6915 Fix communication_error acceptance test
27a5eb7 options[:ca_location] is a string not a symbol. Safer to pull Puppet::SSL::Host.ca_location since it is already set correctly.
c01cac9 (#13204) Don't ignore missing PATH.augnew files
7d27cd2 Use :as_platform in the smf service spec
876785a Use :as_platform in the redhat service spec

2.7.16rc1
===
3a06246 (#8858) Explicitly set SSL peer verification mode.
cd4bee8 Test SSL setup code, not our stubbing of it.

lib/puppet/application/puppetmasterd.rb
require 'puppet'
require 'puppet/application'
require 'puppet/daemon'
require 'puppet/network/server'
require 'puppet/network/http/rack' if Puppet.features.rack?

Puppet::Application.new(:puppetmasterd) do

    should_parse_config

    option("--debug", "-d")
    option("--verbose", "-v")

    # internal option, only to be used by ext/rack/config.ru
    option("--rack")

    option("--compile host",  "-c host") do |arg|
        options[:node] = arg
    end

    option("--logdest DEST",  "-l DEST") do |arg|
        begin
            Puppet::Util::Log.newdestination(arg)
            options[:setdest] = true
        rescue => detail
            if Puppet[:debug]
                puts detail.backtrace
            end
            $stderr.puts detail.to_s
        end
    end

    preinit do
        trap(:INT) do
            $stderr.puts "Cancelling startup"
            exit(0)
        end

        # Create this first-off, so we have ARGV
        @daemon = Puppet::Daemon.new
        @daemon.argv = ARGV.dup
    end

    dispatch do
        if options[:node]
            :compile
        elsif Puppet[:parseonly]
            :parseonly
        else
            :main
        end
    end

    command(:compile) do
        Puppet::Util::Log.newdestination :console
        raise ArgumentError, "Cannot render compiled catalogs without pson support" unless Puppet.features.pson?
        begin
            unless catalog = Puppet::Resource::Catalog.find(options[:node])
                raise "Could not compile catalog for %s" % options[:node] 
            end

            $stdout.puts catalog.render(:pson)
        rescue => detail
            $stderr.puts detail
            exit(30)
        end
        exit(0)
    end

    command(:parseonly) do
        begin
            Puppet::Parser::Interpreter.new.parser(Puppet[:environment])
        rescue => detail
            Puppet.err detail
            exit 1
        end
        exit(0)
    end

    command(:main) do
        require 'etc'
        require 'puppet/file_serving/content'
        require 'puppet/file_serving/metadata'
        require 'puppet/checksum'

        xmlrpc_handlers = [:Status, :FileServer, :Master, :Report, :Filebucket]

        if Puppet[:ca]
            xmlrpc_handlers << :CA
        end

        # Make sure we've got a localhost ssl cert
        Puppet::SSL::Host.localhost

        # And now configure our server to *only* hit the CA for data, because that's
        # all it will have write access to.
        if Puppet::SSL::CertificateAuthority.ca?
            Puppet::SSL::Host.ca_location = :only
        end

        if Process.uid == 0
            begin
                Puppet::Util.chuser
            rescue => detail
                puts detail.backtrace if Puppet[:trace]
                $stderr.puts "Could not change user to %s: %s" % [Puppet[:user], detail]
                exit(39)
            end
        end

        unless options[:rack]
            @daemon.server = Puppet::Network::Server.new(:xmlrpc_handlers => xmlrpc_handlers)
            @daemon.daemonize if Puppet[:daemonize]
        else
            require 'puppet/network/http/rack'
            @app = Puppet::Network::HTTP::Rack.new(:xmlrpc_handlers => xmlrpc_handlers, :protocols => [:rest, :xmlrpc])
        end

        Puppet.notice "Starting Puppet server version %s" % [Puppet.version]

        unless options[:rack]
            @daemon.start
        else
            return @app
        end
    end

    setup do
        # Handle the logging settings.
        if options[:debug] or options[:verbose]
            if options[:debug]
                Puppet::Util::Log.level = :debug
            else
                Puppet::Util::Log.level = :info
            end

            unless Puppet[:daemonize] or options[:rack]
                Puppet::Util::Log.newdestination(:console)
                options[:setdest] = true
            end
        end

        unless options[:setdest]
            Puppet::Util::Log.newdestination(:syslog)
        end

        if Puppet.settings.print_configs?
            exit(Puppet.settings.print_configs ? 0 : 1)
        end

        Puppet.settings.use :main, :puppetmasterd, :ssl

        # Cache our nodes in yaml.  Currently not configurable.
        Puppet::Node.cache_class = :yaml

        # Configure all of the SSL stuff.
        if Puppet::SSL::CertificateAuthority.ca?
            Puppet::SSL::Host.ca_location = :local
            Puppet.settings.use :ca
            Puppet::SSL::CertificateAuthority.instance
        else
            Puppet::SSL::Host.ca_location = :none
        end
    end
end
